### PR TITLE
Add more options and logging capabilities.

### PR DESCRIPTION
**Pull request by Mbucari**
# [Add ChapterTitlesType option](https://github.com/rmcrackan/AudibleApi/commit/5bae3cdbc48983ea3094ee201eaab58bdfb94446)
I discovered an option for how Audible returns chapter info. You can request "Flat" (the old default format) or "Tree", the new default format.

# [Add support to deserialiaze JObject to object](https://github.com/rmcrackan/AudibleApi/commit/0f57b0676be3819e7dc681e537b74b587f54207d)

Just a quality of life method for converting `JObject` to a .NET object without having to convert to string first.

# [Add IsAyce property to Plans](https://github.com/rmcrackan/AudibleApi/commit/715207016805d08cad93735729982c5ddfcd5257)

This is how we detect if a book has any free/plus plans from within Libation. I thought it would be nice to have as a Plan property.

# [Unify Library Api requests](https://github.com/rmcrackan/AudibleApi/commit/1c720284b357466b1225207cb81e8b4d42eedaa4)

I added several more methods for scanning the library, and they all use the `GetLibraryItemsPagesAsync` which returns an `IAsyncEnumerable<Item[]>`. I plan to use these to improve concurrent scan performance in Libation.

# [Add ISourceJson and ReadAsDtoAsync<T>()](https://github.com/rmcrackan/AudibleApi/pull/40/commits/ddd9681eae95caf679fde0cb3030aa5a67983f6b)

There are a couple of things in here:
-  Added `ReadAsDtoAsync<T>()` to read and parse a Dto directly from `HttpContent`. It also takes care of logging errors.
- Added `DtoBase` class with property `SourceJson`. Objects that are `DtoBase` will have their source JObject saved to `DtoBase.SourceJson`.

### CODE REVIEW

**Comment by Mbucari** [comment]
Converting this to draft while I try out another idea for saving source Jason.

File: AudibleApi.Common/Converter.cs [hunk 1]
```diff
@@ -12,22 +17,93 @@ namespace AudibleApi.Common
...
+		public static async Task<T> ReadAsDtoAsync<T>(this HttpContent content) where T : DtoBase<T>
+		{
+			var jobj = await content.ReadAsJObjectAsync();
+			try
+			{
+				return DtoBase<T>.FromJson(jobj);
+			}
+			catch(Exception ex)
+			{
+				Serilog.Log.Logger.Error(ex, $"Error converting {typeof(T).Name}. Full json:\r\n" + jobj.ToString(Formatting.None));
```
Comment (rmcrackan) [inline AudibleApi.Common/Converter.cs:41]: 2 concerns: what if this is friggin huge? what if it contains sensitive info?

File: AudibleApi.Common/_DtoBase.cs [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
 namespace AudibleApi.Common
 {
-	public abstract class DtoBase<T>
+	public abstract class DtoBase
 	{
-		public static T FromJson(string json) => Converter.FromJson<T>(json);
+		public JObject SourceJson { get; set; }
+	}
+
+	public abstract class DtoBase<T> : DtoBase where T : DtoBase<T>
```
Comment (rmcrackan) [inline AudibleApi.Common/_DtoBase.cs:12]: for this and _V10Base, is this `where` correct? I can't tell what it's accomplishing.

File: AudibleApi.Common/Converter.cs (hunk 1 above)
Comment (Mbucari) [reply AudibleApi.Common/Converter.cs:51]: Well, we've been doing it this way.

https://github.com/rmcrackan/AudibleApi/blob/46bb17c70b161bd622f36f34d746de3f22964101/AudibleApi/Api.Library.cs#L321-L331

https://github.com/rmcrackan/AudibleApi/blob/46bb17c70b161bd622f36f34d746de3f22964101/AudibleApi/Api.Library.cs#L249-L261

https://github.com/rmcrackan/AudibleApi/blob/46bb17c70b161bd622f36f34d746de3f22964101/AudibleApi/ApiUnauthenticated.Catalog.cs#L118-L128

https://github.com/rmcrackan/AudibleApi/blob/46bb17c70b161bd622f36f34d746de3f22964101/AudibleApi/ApiUnauthenticated.Content.cs#L21-L30

File: AudibleApi.Common/_DtoBase.cs (hunk 2 above)
Comment (Mbucari) [reply AudibleApi.Common/_DtoBase.cs:12]: Yup. All DTO classes are in the form of `V10Base<T>` or  `DtoBase<T>`

e.g. `Item : DtoBase<Item>`

File: AudibleApi.Common/Converter.cs (hunk 1 above)
Comment (rmcrackan) [reply AudibleApi.Common/Converter.cs:51]: Good point

File: AudibleApi.Common/_DtoBase.cs (hunk 2 above)
Comment (rmcrackan) [reply AudibleApi.Common/_DtoBase.cs:12]: I get that. This `where` seems to be saying, you can only inherit from me if you inherit from me. I can't tell what it's adding.

Comment (Mbucari) [reply AudibleApi.Common/_DtoBase.cs:12]: You're night. Totally superfluous. IDK what I was thinking.

Comment (rmcrackan) [reply AudibleApi.Common/_DtoBase.cs:12]: lol no worries. I felt like I was missing something clever but I couldn't make out what.

File: AudibleApi.Common/Converter.cs (hunk 1 above)
Comment (Mbucari) [reply AudibleApi.Common/Converter.cs:51]: Actually, there is one place where we weren't logging the whole response: GetDownloadLicenseAsync. I'll add an overload for logging level.

Comment (Mbucari) [reply AudibleApi.Common/Converter.cs:51]: nvm. GetDownloadLicenseAsync uses its own special method for parsing to decrypt the voucher. Everywhere ReadAsDtoAsync is used is free of pii.

**Pull request merged**